### PR TITLE
Implement std::str::replacen

### PR DESCRIPTION
**Pull request by knight42**
Replaces first N matches of a pattern with another string.

```
assert_eq!("acaaa".replacen(a, "b", 3), "bcbba")
```

### CODE REVIEW

**Comment by knight42** [comment]
It's rare that @rust-highfive didn't show up :open_mouth:

File: src/libcollections/str.rs [hunk 1]
```diff
@@ -1594,6 +1594,47 @@ impl str {
...
+    /// assert_eq!("foo foo new23 foo", s.replacen(char::is_numeric, "new", 1));
+    /// ```
+    ///
+    /// When the pattern doesn't match:
+    ///
+    /// ```
+    /// let s = "this is old";
+    /// assert_eq!(s, s.replacen("cookie monster", "little lamb", 10));
+    /// ```
+    #[stable(feature = "str_replacen", since = "1.13.0")]
```
Comment (nagisa) [inline src/libcollections/str.rs]: Insta-stable? Where’s the tracking issue which went through FCP for stabilisation?

Comment (knight42) [inline src/libcollections/str.rs]: @nagisa I'm sorry... Should I submit a RFC about this? BTW, where can I know the process of accepting a new method?

Comment (nagisa) [inline src/libcollections/str.rs]: An unstable method with a corresponding tracking issue may be enough.

**Comment by nagisa** [comment]
This PR needs a reviewer, @rust-lang/libs.

File: src/libcollections/str.rs (hunk 1 above)
Comment (knight42) [inline src/libcollections/str.rs]: @nagisa Am I supposed to open the tracking issue by myself?

Comment (nagisa) [inline src/libcollections/str.rs]: I’d suggest to only do that after you get thumbs-up from the libs team for the feature you’re adding. In the mean-time you can use `0` as a placeholder for the attribute.

**Comment by knight42** [comment]
@nagisa Thanks for your advice! :smile:

**Comment by WiSaGaN** [comment]
Seems quite niche to me.

File: src/libcollections/str.rs [hunk 2]
```diff
@@ -1594,6 +1594,49 @@ impl str {
...
+    /// ```
+    #[unstable(feature = "str_replacen",
+               issue = "0",
+               reason = "only need to replace first N matches")]
+    pub fn replacen<'a, P: Pattern<'a>>(&'a self, pat: P, to: &str, count: usize) -> String {
+        // Hope to reduce the times of re-allocation
+        let mut result = String::with_capacity(32);
+        let matched = self.match_indices(pat);
+        let mut last_end = 0;
+        for (_, (start, part)) in (0..count).zip(matched) {
```
Comment (bluss) [inline src/libcollections/str.rs]: Using .take(count) would be the more typical way to limit the number of items here. (I'd understand it right away :D)

File: src/libcollections/str.rs [hunk 3]
```diff
@@ -1594,6 +1594,49 @@ impl str {
...
+            result.push_str(unsafe { self.slice_unchecked(last_end, start) });
```
Comment (bluss) [inline src/libcollections/str.rs]: Usual style is to not use unsafe tightly like this, instead bracket a whole chunk of code where it's needed.

File: src/libcollections/str.rs [hunk 4]
```diff
@@ -1594,6 +1594,49 @@ impl str {
...
+    /// ```
+    /// let s = "this is old";
+    /// assert_eq!(s, s.replacen("cookie monster", "little lamb", 10));
+    /// ```
+    #[unstable(feature = "str_replacen",
+               issue = "0",
+               reason = "only need to replace first N matches")]
+    pub fn replacen<'a, P: Pattern<'a>>(&'a self, pat: P, to: &str, count: usize) -> String {
+        // Hope to reduce the times of re-allocation
+        let mut result = String::with_capacity(32);
```
Comment (bluss) [inline src/libcollections/str.rs]: this number could be based on something from the input

File: src/libcollections/str.rs (hunk 2 above)
Comment (knight42) [inline src/libcollections/str.rs]: Good point!

File: src/libcollections/str.rs (hunk 3 above)
Comment (knight42) [inline src/libcollections/str.rs]: In fact, I copied this line from the implementation of [`str::replace`](https://github.com/rust-lang/rust/pull/36347/files#diff-2544e5e5ce71797344d5ec6acde08b87R1589). Given that `slice_unchecked` is an unsafe function, I think it's Ok.

File: src/libcollections/str.rs (hunk 4 above)
Comment (knight42) [inline src/libcollections/str.rs]: That makes sense, but I guess adding it as another parameter may bother the user?

Or just like the implementation of `str::replace`, we use `let mut result = String::new();` here?

**Comment by alexcrichton** [comment]
Discussed at libs triage today the decision was to merge. @knight42 could you open an issue on rust-lang/rust to track the stabilization of this and update the `#[unstable]` tag reference? I'll be sure to tag it correctly once it's opened.

**Comment by knight42** [comment]
@alexcrichton Great to hear that! :tada:

The `#[unstable]` tag has been updated now.

**Comment by alexcrichton** [comment]
@bors: r+

Thanks!

**Comment by knight42** [comment]
@alexcrichton the failure seems spurious

**Pull request merged**